tangelo/linq/target/target_qiskit.py

```diff
@@ -75,10 +80,57 @@ def simulate_circuit(self, source_circuit: Circuit, return_statevector=False, in
                 initial_state_circuit.initialize(initial_statevector, list(range(n_qubits)))
                 translated_circuit = initial_state_circuit.compose(translated_circuit)
 
-        # Drawing individual shots with the qasm simulator, for noisy simulation or simulating mixed states
-        if self._noise_model or source_circuit.is_mixed_state and not return_statevector:
-            from tangelo.linq.noisy_simulation.noise_models import get_qiskit_noise_model
+        # return_statevector is requested with mid-circuit measurements or
+        # noisy simulation with desired_meas_result so need n_samples with success
+        # so loop shot by shot until (1 success for noiseless simulation, n_shots successes for noisy simulation)
+        if desired_meas_result is not None or (save_mid_circuit_meas and self.n_shots in [None, 1]):
+            n_meas = source_circuit.counts.get("MEASURE", 0)
+            backend = self.AerSimulator(method='statevector')
+            qiskit_noise_model = get_qiskit_noise_model(self._noise_model) if self._noise_model else None
+            translated_circuit = self.qiskit.transpile(translated_circuit, backend)
+            translated_circuit.save_statevector()
 
+            samples = dict()
+            self._current_state = None
+            n_attempts = 0
+            n_success = 0
+            n_shots = self.n_shots if self.n_shots is not None else -1
+
+            # Permit 0.1% probability events
+            max_attempts = 1000*abs(n_shots)
+
+            while self._current_state is None and n_attempts < max_attempts and n_success != n_shots:
+                sim_results = backend.run(translated_circuit, noise_model=qiskit_noise_model, shots=1).result()
+                current_state = sim_results.get_statevector(translated_circuit)
+                measure = next(iter(self.qiskit.result.marginal_counts(sim_results, indices=list(range(n_meas))).get_counts()))[::-1]
+                if measure == desired_meas_result:
+                    if not self._noise_model:
+                        self._current_state = current_state
+                        if self.n_shots is not None:
+                            self.all_frequencies = {measure + state[::-1]: count for state, count in current_state.sample_counts(self.n_shots).items()}
+                        else:
+                            freqs = self._statevector_to_frequencies(np.array(current_state))
+                            self.all_frequencies = {measure + meas: val for meas, val in freqs.items()}
+                        return (self.all_frequencies, np.array(self._current_state)) if return_statevector else (self.all_frequencies, None)
+                    else:
+                        (sample, _) = current_state.measure()
+                        bitstr = measure + sample[::-1]
+                        samples[bitstr] = samples.get(bitstr, 0) + 1
+                        n_success += 1
+                elif desired_meas_result is None:
```

Comment:
It feels like control flow here may be more complicated that it could be. I'm not sure I understand what's going on. If there are no desired measurement, then why are we inside this complicated while loop in the first place ?
Yet it's inside the branch led by L86 so that means it cannot be `None` anyway ?

I suggest you separate and organize the different cases clearly because `linq` is becoming tough to handle. You are free to create more functions if that helps. Skip lines, use comments, whatever: let's try to make this complicated logic as accessible as possible. It's ok if some lines are redundant, the goal is to end up with a code that's readable and easier to maintain.